test/databases/transaction_database.cpp

```diff
@@ -38,51 +38,502 @@ struct transaction_database_directory_setup_fixture
     {
         test::clear_path(DIRECTORY);
     }
+
+    ~transaction_database_directory_setup_fixture()
+    {
+        test::clear_path(DIRECTORY);
+    }
 };
 
-BOOST_FIXTURE_TEST_SUITE(database_tests, transaction_database_directory_setup_fixture)
+BOOST_FIXTURE_TEST_SUITE(transaction_database_tests, transaction_database_directory_setup_fixture)
+
+BOOST_AUTO_TEST_CASE(transaction_database__store__single_transactions__success)
+{
+    transaction tx1;
+    data_chunk wire_tx1;
+    BOOST_REQUIRE(decode_base16(wire_tx1, TRANSACTION1));
+    BOOST_REQUIRE(tx1.from_data(wire_tx1));
+
+    transaction tx2;
+    data_chunk wire_tx2;
+    BOOST_REQUIRE(decode_base16(wire_tx2, TRANSACTION2));
+    BOOST_REQUIRE(tx2.from_data(wire_tx2));
+
+    const auto path = DIRECTORY "/tx_table";
+    test::create(path);
+    transaction_database db(path, 1000, 50, 0);
+    BOOST_REQUIRE(db.create());
+
+    const auto hash1 = tx1.hash();
+    BOOST_REQUIRE(!db.get(hash1));
+
+    const auto hash2 = tx2.hash();
+    BOOST_REQUIRE(!db.get(hash2));
+
+    // Setup end
+    
+    db.store({tx1}, 100, 110);
+
+    const auto result1 = db.get(hash1);
+    BOOST_REQUIRE(result1);
+    BOOST_REQUIRE(result1.transaction().hash() == hash1);
+
+    db.store({tx2}, 200, 220);
```

Comment:
I went with the list as there is no method in the class that takes a single confirmed transaction with height/fork and MTP.

Should we add such a method?

Here's the three we have for now

```
    /// Store a transaction not associated with a block.
    bool store(const chain::transaction& tx, uint32_t forks);

    /// Store a set of transactions associated with an unconfirmed block.
    bool store(const chain::transaction::list& transactions);

    /// Store a set of transactions associated with a confirmed block.
    bool store(const chain::transaction::list& transactions, size_t height,
        uint32_t median_time_past);
```

Reply:
No, since there is no scenario where we would have a single confirmation independent of the set of others in a block. So you could just just change the test to:

```cpp
db.store(tx1, 100);
...
db.store(tx2, 200)
```

Reply:
Actually there is an envisionable scenario where would would pass individual confirmed transactions independently - specifically where we are fanning out to write them concurrently. However this is not a current optimization since we are writing blocks in parallel during catch-up and we are not writing a block of confirmed transactions at all when confirming a self-mined block or compact block. This made me realize that this method is actually dead code:
```cpp
bool store(const chain::transaction::list& transactions, size_t height,
        uint32_t median_time_past);
```
So I have removed it in a pending pull request.

Reply:
Got it. Making the appropriate changes.

Reply:
@evoskuil - just thinking about what you said there about removing the method. Once removed, there is no way to "store a confirmed transaction". We will need to store a list of transactions and and then call confirm on the list.

I am facing this situation while trying to wrap up a test case for unconfirm. With the above mentioned method removed, I have no way left to store a confirmed transaction. I have to store it first, then confirm it as part of the setup.  Only then can I call unconfirm to test that unconfirm does the necessary things.

This would result in two mutexes being taken, first the root_mutex of the hash table and then the transaction database's metadata_mutex. I don't know yet if it is good or bad, just sharing the observation.

Reply:
The number of locks taken is not the important performance consideration given the overall design. The ability to download and store blocks in parallel is a greater optimization, and this implies that blocks are not yet validated. As such they cannot be confirmed when stored. Even when operating at the top of the chain it is generally the case that txs are stored unvalidated (via the tx pool or via request into that pool from a gap in a compact block) and later confirmed.

Reply:
Agree with not supporting a method only for tests. I was wondering if we might need it anyway later on, but you clarified why we won't need it.

I'll change the test to store and then confirm.